### PR TITLE
Initial vmap frontend API

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40455 Add batching rule for unsqueeze and transpose
* **#40172 Initial vmap frontend API**
* #40171 Add invariants check to BatchedTensorImpl

This PR introduces the initial vmap frontend API. It has the following
limitations that we can resolve in the future:
- the inputs must be a flat list of tensors
- the outputs must be a flat list of tensors
- in_dims = 0 (so we always vmap over dim 0 of input tensors)
- out_dims = 0 (so the returned tensors have their vmap dim appear at
dim 0)
- Coverage limited to operations that have batching rules implemented
(torch.mul, torch.sum, torch.expand).

There are some other semantic limitations (like not being able to handle
mutation, aside from pytorch operations that perform mutation) that will
be documented in the future.

I wanted to introduce the API before adding a slow fallback for the
coverage so that we can test future batching rules (and coverage) via
the python API to avoid verbosity in C++-land.

The way vmap works is that `vmap(func)(inputs)` wraps all Tensor inputs
to be batched in BatchedTensors, sends those into func, and then unwraps
the output BatchedTensors. Operations on BatchedTensors perform the batched
operations that the user is asking for. When performing nested vmaps,
each nested vmap adds a batch dimension upon entry and removes a batch
dimension on exit.

Coming up in the near future:
- Support for non-zero in_dims and out_dims
- docstring for vmap
- slow fallback for operators that do not have a batching rule
implemented.

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22102076](https://our.internmc.facebook.com/intern/diff/D22102076)

### CODE REVIEW

File: torch/vmap_internals.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import torch
```
Comment (ezyang) [inline torch/vmap_internals.py:1]: If this is internals it should have a leading underscore

File: aten/src/ATen/BatchingRegistrations.cpp [hunk 2]
```diff
@@ -112,6 +112,8 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   // Tensor wrapper, it only has one dispatch key (Batched) on it. The resolution
   // here is to just directly call the underlying implementation.
   m.impl("size.int", static_cast<int64_t (*)(const Tensor&, int64_t)>(native::size));
+  m.impl("_add_batch_dim", native::_add_batch_dim);
```
Comment (ezyang) [inline aten/src/ATen/BatchingRegistrations.cpp:115]: Shouldn't these functions also work on non-batched tensors too?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -6642,6 +6642,12 @@
   variants: function, method
   device_guard: False
 
+- func: _add_batch_dim(Tensor self, int batch_dim, int level) -> Tensor
+  use_c10_dispatcher: full
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:6646]: Make these function only?

Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:6646]: Is the intention to eventually delete these functions when we have more vmap functionality? If not, why do we need these functions--does it make sense to make these functions part of the *public* API for batched tensors?

Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6646]: iirc the default for native_functions.yaml is function only. I can the `variants: function` line to make that explicit, though

File: aten/src/ATen/BatchingRegistrations.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/BatchingRegistrations.cpp:115]: They do work on non-batched tensors. They're here because BatchedTensor needs a registration for them, otherwise, `_add_batch_dim`/`_remove_batch_dim` only work on regular Tensors

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6646]: > Is the intention to eventually delete these functions when we have more vmap functionality? If not, why do we need these functions--does it make sense to make these functions part of the public API for batched tensors?

These functions are part of the pytorch developer API for batched tensors. However, pytorch users shouldn't be exposed to batched tensors, so they should never see these functions.

I was not planning on deleting these functions, but I think it makes sense to put them in a namespace in the future (so they do not appear as functions on torch)

**Comment by seemethere** [comment]
Hey @zou3519 just wanted to check in on this PR, we will be cutting our release branch sometime tomorrow so would love to have this merged before that.

**Comment by zou3519** [comment]
@seemethere this commit shouldn't go into the release. If it ends up going into the release branch then the plan is to push a change to remove the api from the release branch

File: torch/vmap_internals.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            continue
+        raise ValueError(ELEMENT_MUST_BE_TENSOR.format(fn=fn_name, out=type(output), idx=idx))
+
+# This is the global tracker for how many nested vmaps we are currently inside.
+VMAP_LEVEL = 0
+
+# vmap(func)(inputs) wraps all Tensor inputs to be batched in BatchedTensors,
+# sends those into func, and then unwraps the output BatchedTensors. Operations
+# on BatchedTensors perform the batched operations that the user is asking for.
+def vmap(func, in_dims=0, out_dims=0):
```
Comment (ezyang) [inline torch/vmap_internals.py:64]: Put a docblock here?

File: torch/vmap_internals.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+def _validate_inputs_and_get_batch_size(args, fn_name):
+    if len(args) == 0:
+        raise ValueError(NO_INPUTS.format(fn=fn_name))
+    return _validate_and_get_batch_size(args)
+
+# Undos the batching (and any batch dimensions) associated with the `vmap_level`.
+def _unwrap_batched(batched_outputs, vmap_level, batch_size):
+    if isinstance(batched_outputs, Tensor):
+        return torch._remove_batch_dim(batched_outputs, vmap_level, batch_size, 0)  # type: ignore
```
Comment (ezyang) [inline torch/vmap_internals.py:43]: OOC, why do you need to `type: ignore` here? We should have accurate type annotations for these functions!

File: test/test_vmap.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+from torch.testing._internal.common_utils import TestCase, run_tests
+import torch
+from torch import vmap
```
Comment (ezyang) [inline test/test_vmap.py:3]: Do we have plans for a more structured testing strategy? In principle you have to test vmap for *every* operator and it would be much better if we can incorporate vmap testing as part of, e.g., `common_methods`

File: torch/vmap_internals.py (hunk 4 above)
Comment (zou3519) [reply torch/vmap_internals.py:64]: The docstring containing the vmap user documentation is coming in a future update, after I implement the `in_dims` and `out_dims` functionality, so that it'll be correct. I can put something minimal here for now if you think that would be good

File: test/test_vmap.py (hunk 6 above)
Comment (zou3519) [reply test/test_vmap.py:3]: My currently testing strategy for operators is to construct a new table for vmap edge cases for every operator. `common_method_invocations` doesn't contain enough information for vmap tests.

I've considered extending `common_method_invocations` but that idea did not go very far

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:6646]: Probably would be a good idea to documented the specific situations where these functions are intended to be used, so it's easier for us to assess if other future users are using them inappropriately.

File: torch/vmap_internals.py (hunk 5 above)
Comment (zou3519) [reply torch/vmap_internals.py:43]: I think the answer is mypy doesn't like that we're calling methods on torch while we `import torch` (this file gets imported as a part of the `import torch` process) but I didn't dig very far into the error. I'll dig a little and then put a comment to that effect here if my hypothesis is supported

File: test/test_vmap.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        self.assertEqual(output, x * y)
+
+        output = vmap(vmap(vmap(torch.mul)))(x, y)
+        self.assertEqual(output, x * y)
+
+    def test_nested_with_different_map_dim(self):
+        x = torch.randn(2, 3)
+        y = torch.randn(5, 3)
+        output = vmap(lambda x: vmap(lambda y: x * y)(y))(x)
+        self.assertEqual(output, x.view(2, 1, 3) * y)
```
Comment (ezyang) [inline test/test_vmap.py:79]: I did a double take here because I'm expecting the size to be `(2, 5, 3)`

File: test/test_vmap.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            vmap(foo)(x)
+
+    def test_unsupported_inplace_op_err_msg(self):
+        def foo(x):
+            return x.cos_()
+
+        x = torch.randn(3)
+        # TODO(rzou): Yeah, this error message is pretty bad because the
+        # dispatcher's fallback mechanism doesn't work for ops that don't support
+        # boxing. Fix the error message at some point.
```
Comment (ezyang) [inline test/test_vmap.py:106]: If you have ideas for how to make the default error message better, patches welcome :)

**Review by ezyang — APPROVED** [review]

**Review by ezyang — APPROVED** [review]

File: test/test_vmap.py (hunk 7 above)
Comment (zou3519) [reply test/test_vmap.py:79]: Yeah, I can see that. I'll add an explicit check for the size so that it is obvious that the result should be 2, 5, 3

File: torch/vmap_internals.py (hunk 5 above)
Comment (zou3519) [reply torch/vmap_internals.py:43]: Update: I spent some time looking into this. mypy is ignoring all functions on torch that begin with an underscore. I don't know why this is (debugging the type checker is interestingly difficult) but I filed an issue at #40397. I'll just link to that issue from an in-line comment here.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6627]: Documented.

File: torch/vmap_internals.py (hunk 5 above)
Comment (ezyang) [reply torch/vmap_internals.py]: One way that might work to workaorund this problem is to say `torch._C._VariableFunctions._remove_batch_dim` instead. (Note that this workaround doesn't work if you're TorchScripting)
